src/integrity-checker.js

```diff
@@ -117,12 +199,38 @@ export default class InstallationIntegrityChecker {
       };
     }
 
-    const actual = this._generateIntegrityHash(normalizedLockSource, patterns, flags);
-    const expected = (await fs.readFile(loc.locationPath)).trim();
+    const actual = await this._generateIntegrityFile(
+      lockfile,
+      patterns,
+      Object.assign({}, {checkFiles: false}, flags), // don't generate files when checking, we check the files below
+      loc.locationFolder);
+    const expectedRaw = await fs.readFile(loc.locationPath);
+    let expected: IntegrityFile;
+    try {
+      expected = JSON.parse(expectedRaw);
+    } catch (e) {
+      // ignore JSON parsing for legacy text integrity files compatibility
+    }
+    let integrityMatches;
+    if (expected) {
+      integrityMatches = this._compareIntegrityFiles(actual, expected);
+      if (flags.checkFiles && expected.files.length > 0) {
+        // TODO we may want to optimise this check by checking only for package.json files on very large trees
```

Comment:
Or maybe on random-but-reproductible selected files? For example one file every two entries?

Another idea: register files and directories in two separate lists, and map them with their mtime. During the integrity check, just compare the directories mtime on the filesystem with those they should have. If they don't match, it's that the directory has been tempered with, and you can check its content to see if a file has been deleted. This would drastically decrease the number of I/O checks for the most common case.